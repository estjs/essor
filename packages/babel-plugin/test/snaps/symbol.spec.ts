import { getTransform } from './transform';

const transformCode = getTransform('symbol');
describe('transform symbol', () => {
  it('should work with basic types', () => {
    const list = [false, true, 0, 1, '1', `${1}23`, null, undefined, Number.NaN];

    for (const item of list) {
      expect(transformCode(`const $a = ${item}; $a = 1`)).toMatchSnapshot();
    }
  });

  it('should work with array', () => {
    const list = ['[1,2,3]', '[1,{a:1},3,4]', '[1,2,3,4,5]'];
    for (const item of list) {
      expect(transformCode(`const $a = ${item}; $a = 2`)).toMatchSnapshot();
    }
  });
  it('should work with object', () => {
    const list = ['{a:1}', '{a:1,b:2}', '{a:1,b:{c:1}}'];
    for (const item of list) {
      // if has assignment, transform signal
      expect(transformCode(`const $a = ${item}; $a = 2`)).toMatchSnapshot();
      expect(transformCode(`const $a = ${item}; `)).toMatchSnapshot();
    }
  });

  it('should work with const function transform computed', () => {
    const list = ['() => 1', '(a) => a + 1'];
    for (const item of list) {
      expect(transformCode(`const $a = ${item}; $a = 2`)).toMatchSnapshot();
    }
  });

  it('should work with let function transform computed', () => {
    const list = ['() => 1', '(a) => a + 1'];
    for (const item of list) {
      expect(transformCode(`let $a = ${item}; $a = 2`)).toMatchSnapshot();
    }
  });
  it('should work with object pattern', () => {
    const input = `
      const {$a,b} = {$a:1,b:2};
      const [$c,d] = [1,2];
      const [{$d,e}] = [{$d:1,e:2}];
      console.log($a,b);
      console.log($c,d);
      console.log($d,e);
    `;
    expect(transformCode(input)).toMatchSnapshot();
  });
  it('should work with object pattern alias', () => {
    const input = `
      const {a:$a,b} = {a:1,b:2};
      const [{d:$d}] = [{d:1,e:2}];
      console.log($a,b);
      console.log($d,);
  `;
    expect(transformCode(input)).toMatchSnapshot();
  });

  it('should work with import value', () => {
    const input = `
      import {$a} from 'a';
      import {b} from 'b';
      console.log($a,b);
  `;

    expect(transformCode(input)).toMatchSnapshot();
  });

  it('should work with array index ', () => {
    const input = `
      const $a = [1];
      console.log($a[0]);
  `;

    expect(transformCode(input)).toMatchSnapshot();
  });
  it('should not work with add value symbol', () => {
    const input = `
      const $a = [1];
      $a.value.push(2);
  `;
    expect(transformCode(input)).toMatchSnapshot();
  });
});

const transformCustomSymbol = getTransform('symbol', { symbol: '__' });
describe('transform custom symbol', () => {
  it('should work with basic types', () => {
    const list = [false, true, 0, 1, '1', `${1}23`, null, undefined, Number.NaN];
    for (const item of list) {
      expect(transformCustomSymbol(`const __a = ${item}; __a = 1`)).toMatchSnapshot();
    }
  });
});
