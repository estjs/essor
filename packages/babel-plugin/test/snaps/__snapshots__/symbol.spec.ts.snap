// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform custom symbol > should work with basic types 1`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(false);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 2`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(true);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 3`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(0);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 4`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(1);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 5`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(1);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 6`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(123);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 7`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(null);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 8`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(undefined);
__a.value = 1;"
`;

exports[`transform custom symbol > should work with basic types 9`] = `
"import { signal as _signal$ } from "essor";
const __a = _signal$(NaN);
__a.value = 1;"
`;

exports[`transform symbol > should not work with add value symbol 1`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$([1]);
$a.value.push(2);"
`;

exports[`transform symbol > should work with array 1`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$([1, 2, 3]);
$a.value = 2;"
`;

exports[`transform symbol > should work with array 2`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$([1, {
  a: 1
}, 3, 4]);
$a.value = 2;"
`;

exports[`transform symbol > should work with array 3`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$([1, 2, 3, 4, 5]);
$a.value = 2;"
`;

exports[`transform symbol > should work with array index  1`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$([1]);
console.log($a.value[0]);"
`;

exports[`transform symbol > should work with basic types 1`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(false);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 2`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(true);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 3`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(0);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 4`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(1);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 5`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(1);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 6`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(123);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 7`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(null);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 8`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(undefined);
$a.value = 1;"
`;

exports[`transform symbol > should work with basic types 9`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$(NaN);
$a.value = 1;"
`;

exports[`transform symbol > should work with const function transform computed 1`] = `
"import { computed as _computed$ } from "essor";
const $a = _computed$(() => 1);
$a.value = 2;"
`;

exports[`transform symbol > should work with const function transform computed 2`] = `
"import { computed as _computed$ } from "essor";
const $a = _computed$(a => a + 1);
$a.value = 2;"
`;

exports[`transform symbol > should work with import value 1`] = `
"import { $a } from 'a';
import { b } from 'b';
console.log($a.value, b);"
`;

exports[`transform symbol > should work with let function transform computed 1`] = `
"import { signal as _signal$ } from "essor";
let $a = _signal$(() => 1);
$a.value = 2;"
`;

exports[`transform symbol > should work with let function transform computed 2`] = `
"import { signal as _signal$ } from "essor";
let $a = _signal$(a => a + 1);
$a.value = 2;"
`;

exports[`transform symbol > should work with object 1`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1
});
$a.value = 2;"
`;

exports[`transform symbol > should work with object 2`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1
});"
`;

exports[`transform symbol > should work with object 3`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1,
  b: 2
});
$a.value = 2;"
`;

exports[`transform symbol > should work with object 4`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1,
  b: 2
});"
`;

exports[`transform symbol > should work with object 5`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1,
  b: {
    c: 1
  }
});
$a.value = 2;"
`;

exports[`transform symbol > should work with object 6`] = `
"import { signal as _signal$ } from "essor";
const $a = _signal$({
  a: 1,
  b: {
    c: 1
  }
});"
`;

exports[`transform symbol > should work with object pattern 1`] = `
"const {
  $a,
  b
} = {
  $a: 1,
  b: 2
};
const [$c, d] = [1, 2];
const [{
  $d,
  e
}] = [{
  $d: 1,
  e: 2
}];
console.log($a.value, b);
console.log($c.value, d);
console.log($d.value, e);"
`;

exports[`transform symbol > should work with object pattern alias 1`] = `
"const {
  a: $a,
  b
} = {
  a: 1,
  b: 2
};
const [{
  d: $d
}] = [{
  d: 1,
  e: 2
}];
console.log($a.value, b);
console.log($d.value);"
`;
